homeassistant/components/imap/coordinator.py

```diff
@@ -390,14 +418,25 @@ async def _async_wait_push_loop(self) -> None:
                 await asyncio.sleep(BACKOFF_TIME)
                 continue
             else:
+                self.auth_errors = 0
                 self.async_set_updated_data(number_of_messages)
+                if self.config_entry.state == ConfigEntryState.SETUP_ERROR:
+                    self.config_entry.async_set_state(
```

Comment:
This isn't safe to call outside of the config entry class.  I should have made it a private function.

Reply:
Oh no, I missed the use of it in #89759

Reply:
The `ConfigEntry` class doesn't consider the state being changed by anything but itself so it may work until it explodes one day because the design assumes that can't happen.

Reply:
You suggest not using the set state calls here?

Reply:
Ideally we come up with a way to trigger the config entry to flip state without calling it directly so the flow of changing it it entirely handled by the config entry class itself.

Reply:
I think it could be refactored so we do a single poll at startup and still `raise ConfigEntryNotReady` and than move to the push pattern as soon as we know everything is good. That would ensure we never change the state unexpectedly out from the ConfigEntry class.

Apologizes for not noticing this sooner.

Reply:
Some more backstory

By overriding the state to setup_retry instead of loaded the reload code is likely to behave badly because it will make the wrong assumptions about the config entry. It being loaded.  Keep in mind that the user can reload the config entry at any time from the UI.

Reply:
I have refactored all async_set_states it. It was used to indicate an invalid folder. Usually this is something static. The triggering reauth mechanism stays, allowing 2 extra tries befor triggering a reauth to make it more robust.